code/modules/holodeck/computer.dm

```diff
@@ -110,207 +126,267 @@
-			var/program_to_load = text2path(params["type"])
-			if(!ispath(program_to_load))
-				return FALSE
-			var/valid = FALSE
+			var/program_to_load = params["id"]
+
 			var/list/checked = program_cache.Copy()
-			if(obj_flags & EMAGGED)
+			if (obj_flags & EMAGGED)
 				checked |= emag_programs
-			for(var/prog in checked)
-				var/list/P = prog
-				if(P["type"] == program_to_load)
+			var/valid = FALSE //dont tell security about this
+
+			for (var/prog in checked)//checks if program_to_load is any one of the loadable programs, if it isnt then it rejects it
+				var/list/check_list = prog
+				if (check_list["id"] == program_to_load)
 					valid = TRUE
 					break
-			if(!valid)
+			if (!valid)
 				return FALSE
-
-			var/area/A = locate(program_to_load) in GLOB.sortedAreas
-			if(A)
-				load_program(A)
+			//load the map_template that program_to_load represents
+			if(program_to_load)
+				load_program(program_to_load)
 		if("safety")
 			if((obj_flags & EMAGGED) && program)
 				emergency_shutdown()
-			nerf(obj_flags & EMAGGED)
+			nerf(obj_flags & EMAGGED,FALSE)
 			obj_flags ^= EMAGGED
-			say("Safeties restored. Restarting...")
+			say("Safeties reset. Restarting...")
+
+///this is what makes the holodeck not spawn anything on broken tiles (space and non engine plating / non holofloors)
+/datum/map_template/holodeck/update_blacklist(turf/placement)
+	. = list()
+	for (var/_turf in get_affected_turfs(placement))
+		var/turf/possible_blacklist = _turf
+		if (!istype(possible_blacklist, /turf/open/floor/holofloor))
+			if (istype(possible_blacklist, /turf/open/floor/engine))
+				continue
+			. += possible_blacklist
+	return .
+
+///loads the template whose id string it was given ("offline_program" loads datum/map_template/holodeck/offline)
+/obj/machinery/computer/holodeck/proc/load_program(map_id, force = FALSE, add_delay = TRUE)
+	if (program == map_id)
+		return
+
+	if (!is_operational)//load_program is called once with a timer (in toggle_power) we dont want this to load anything if its off
+		map_id = offline_program
+		force = TRUE
+
+	if (current_cd > world.time && !force)
+		say("ERROR. Recalibrating projection apparatus.")
+		return
+
+	if (add_delay)
+		current_cd = world.time + HOLODECK_CD
+		if(damaged)
+			current_cd += HOLODECK_DMG_CD
+			damaged = FALSE
+	active = (map_id != offline_program)
+	use_power = active + IDLE_POWER_USE
+	program = map_id
+
+	//clear the items from the previous program
+	for (var/_item in spawned)
+		var/obj/holo_item = _item
+		derez(holo_item)
+
+	for (var/_effect in effects)
+		var/obj/effect/holodeck_effect/holo_effect = _effect
+		effects -= holo_effect
+		holo_effect.deactivate(src)
+
+	//makes sure that any time a holoturf is inside a baseturf list (e.g. if someone put a wall over it) its set to the OFFLINE turf
+	//so that you cant bring turfs from previous programs into other ones (like putting the plasma burn turf into lounge for example)
+	for (var/turf/closed/holo_turf in linked)
```

Comment:
Does `link` contain a different type than this one?

Reply:
yes, linked contains every non area atom inside the holodeck room